compiler: Change tile use in DeviceAcczier to allow multiple tile sizes

gpc1064 opened the pull request:
A proposal to allow the use of multiple tile sizes in the same operator, through the _part-tile_ flag.
Passing multiple tuples to the _part-tile_ flag is already supported, but only the first tuple is applied in the __make_partree_ method of _DeviceAcczier_ class.

The changes consist of supplying the nesting index in the _make_parallel_ method of _PragmaShmTransformer_ to the __make_partree_ method and using the tile size with the respective index.

Cases with different tile and nesting sizes are handled. Given `n` tile sizes for an operator with `m` nests:

If `n = m`: The i-th tile size is assigned to the i-th nesting.
If `n > m`: The first `m` tile sizes are assigned to the `m` nests (i-th tile size -> i-th nesting), and the remaining `n-m` is not used.
If `n < m`: The `n` tile sizes are assigned to the first `n` nestings, and the remaining `m-n` nestings replicate the last tile size.

FabioLuporini commented on devito/passes/iet/languages/openacc.py:
You may now drop this comment

georgebisbas reviewed (changes requested):
Thanks for this! As a first thing to improve, before the rest, could you add tests to this PR?

FabioLuporini commented on devito/passes/iet/languages/openacc.py:
how about simply `n = ...` ?

FabioLuporini commented on devito/passes/iet/languages/openacc.py:
we prefer explicit if-then-else:

```
if ...:
    idx = ...
else:
    idx = ...
```

FabioLuporini replied:
and in practice this is just 

```
index = min(index, par_tile_length-1)
```

gpc1064 replied:
Which one is more suitable?
```
tile = self.par_tile[min(index, len(self.par_tile) - 1)]
```

```
idx = min(index, len(self.par_tile) - 1)
tile = self.par_tile[idx]
```

FabioLuporini replied:
I'd say the latter for homogeneity with the rest of the codebase.

FabioLuporini commented on devito/passes/iet/parpragma.py:
```diff
@@ -388,14 +388,14 @@ def _make_nested_partree(self, partree):
 
             # Outer parallelism
-            root, partree = self._make_partree(candidates)
+            root, partree = self._make_partree(candidates, None, i)
```
instead of passing None:

```
self._make_partree(candidates, index=i)
```

FabioLuporini reviewed (changes requested):
This PR needs to be equipped with tests, somewhere here: https://github.com/devitocodes/devito/blob/master/tests/test_dle.py#L265

and at least one in `test_gpu_openacc.py` as well

I've also left a batch of small comments

gpc1064 commented:
> Thanks for this! As a first thing to improve, before the rest, could you add tests to this PR?

> This PR needs to be equipped with tests, somewhere here: https://github.com/devitocodes/devito/blob/master/tests/test_dle.py#L265
> 
> and at least one in `test_gpu_openacc.py` as well
> 
> I've also left a batch of small comments

Added [test_openacc_structure](https://github.com/gpc1064/devito/blob/multiple_par_tile/tests/test_dle.py#L329) in `test_dle.py::TestBlockingParTile` to ensure correct blocking while used with OpenACC's tile.

Added [test_openacc_multi_tile_structure](https://github.com/gpc1064/devito/blob/multiple_par_tile/tests/test_dle.py#L350) in `test_dle.py::TestBlockingParTile` to ensure correct blocking and applying multiple tile sizes with OpenACC.

Added [test_multiple_tile_sizes](https://github.com/gpc1064/devito/blob/multiple_par_tile/tests/test_gpu_openacc.py#L116) in `test_gpu_openacc.py::TestCodeGeneration` ensuring that multiple tile sizes are applied correctly across nests.

gpc1064 commented:
Requested code changes also committed

FabioLuporini commented on tests/test_dle.py:
```diff
@@ -318,6 +318,60 @@ def test_structure_2p5D(self):
+        (((32, 4, 4), 1, 'tag0'), ((4, 4, 32), (4, 4, 32))),
+        ((((32, 4, 8), 1, 'tag0'), ((32, 8, 4), 2)), ((8, 4, 32), (4, 8, 32))),
+    ])
+    def test_openacc_structure(self, par_tile, expected):
```
Sorry, I made a mistake; clearly these tests belong to `test_gpu_openacc`, not to `test_dle` which constitutes a platform-independent set of tests

FabioLuporini replied:
(By moving them there you'll have to rename them as well -- dropping "openacc" for example)

FabioLuporini commented on tests/test_dle.py:
```diff
@@ -318,6 +318,60 @@ def test_structure_2p5D(self):
+            assert all(i.step == j for i, j in zip(iters, v))
+
+
+    def test_openacc_multi_tile_structure(self):
```
how is this test conceptually different than the one above?

gpc1064 replied:
Indeed the first one was kind of unnecessary. I've dropped it and kept the second one, which is more relevant

FabioLuporini reviewed (approved):
OK I think this is now good to go. Enabling tests...
